### DISCUSSION TITLE
Collaboration/Conversation Inquiry: Evaluation On bis-file-formats; Thoughts On Switching Bases If Applicable In Future

### DISCUSSION THREAD

**Discussion by FlipperPlz** [jetelain/ArmaRealMap#111] [General]
Hey Jetelain! 👋 

I've been interested in your work for a long time since our two spheres collide (I work on a library called BisUtils, aimed to be a "perfect" framework for all things bis). I wanted to touch base with you and hear what it is you use the most from bis-file-formats. After I get my library finished and out in beta, I plan on forking some of your applications from bis-file-formats to BisUtils as a sample project. I have been working with Bis and its many formats and niche functionality. My discord is @flipperplz if you ever want to get in contact with me.

Thanks for making opensource software, you are empowering young developers aswell as the modding community centered around Bohemia's games. Keep up the great work!

**Comment by jetelain** [comment]
Hi,

This could be a great idea, bis-file-formats need rework but does the job meanwhile. It would great to migrate to BisUtils.

In GameRealisticMap I use a lot of features:
- WRP
  - Read (to edit and write again)
  - Write (to generate files)
- PBO 
  - Read (to be able to unpack references files)
  - Write (to generate a mod without the need of PboProject)
- PAA
  - Read, with conversion to WPF or ImageSharp (depending on the context)
  - Writing would be great, but it seems nobody figured out how to compute the "average color". I still use Image2Paa for writing.
- Config (config.bin, rvmat)
  - Read
  - Writing
  - I'm using CfgConvert to binarize config, but a builtin tool could be also great (to have better error messages)
- P3D ODOL
  - Read
    - ModelInfo: BoundingCenter (required to generate WRP), Class, etc.   
    - LODs: Properties (such as slopelandcontact), Faces (to detect model real size, "trunk" position), materials, textures, etc.

In the future, I may need to write P3D MLOD for templating purpose (procedural buildings).

**Reply by FlipperPlz** [reply]
In the ongoing pursuit of stability and robustness, my recent efforts have been concentrated on internal parsing mechanisms, with minimal reliance on external libraries. To this end, I've only incorporated Microsoft's LoggerAbstractions package for ILogger.

For your reference, here are some key achievements:

- Models have been implemented for P3Ds, with a focus on reading and writing odol/mlod files using a unified class structure. The implementation of these models, however, is a work in progress, and the separation of their binarized counterparts remains to be fully realized.
- The development of functionalities for Binarize/Debinarize operations has been undertaken for `rapify` files, alongside an in-house parser built upon the `LangAssembler` library that I just recently moved to its own repository. Additionally, a preprocessor has been crafted to manage macros and includes within these files.
- Models pertaining to `PAA`s are also in development, although these efforts are ongoing.
- I have achieved a comprehensive and accurate method for `PBO` reading and writing, with an option to load them in a flattened format to streamline processes that do not necessitate the directory layout.

I look forward to the possibility of further discussions and collaborations on these matters, and I will keep you updated on the progress of the other projects needed.

Best regards, Ryann
